docs/ballooning.md

```diff
@@ -0,0 +1,257 @@
+guest driver.
+
+Please note that even in the case where the driver is not working properly,
+the balloon will never leak memory from one Firecracker process to another,
+nor can a guest within Firecracker access information in memory outside its
+own guest memory. In other words, memory cannot leak in or out of Firecracker
+if the driver becomes corrupted. This is guaranteed by the fact that the page
+frame numbers coming from the driver are checked to be inside the guest
+memory, then `madvise`d with the `MADV_DONTNEED` flag, which breaks the
+mappings between host physical memory (where the information is ultimately
+stored) and Firecracker virtual memory, which is what Firecracker uses to
+build the guest memory. On subsequent accesses on previously `madvise`d
+memory addresses, the memory is zeroed. Furthermore, the guest memory is
+`mmap`ped with the `MAP_PRIVATE` and `MAP_ANONYMOUS` flags, which ensure that
+even if a Firecracker yields some information through an inflate and that
+same physical page containing the information is mapped onto another
+Firecracker process, reads on that address space will see zeroes.
+
+## Prerequisites
+
+To support memory ballooning, you must use a kernel that has the memory
+ballooning driver installed (on Linux 4.14.7, the relevant settings are
+`CONFIG_MEMORY_BALLOON=y`, `CONFIG_VIRTIO_BALLOON=y`). Other than that, only
+the requirements mentioned in the `getting-started` document are needed.
+
+## Installing the balloon device
+
+In order to use a balloon device, you must install it during virtual machine
+setup (i.e. before starting the virtual machine). This can be done either
+through a PUT request on "/balloon" or by inserting the balloon into the JSON
+configuration file given as a command line argument to the Firecracker process.
+
+Here is an example command on how to install the balloon through the API:
+
+```
+socket_location=...
+amount_mb=...
+must_tell_host=...
+deflate_on_oom=...
+polling_interval=...
+
+curl --unix-socket $socket_location -i \
+    -X PUT 'http://localhost/balloon' \
+    -H 'Accept: application/json' \
+    -H 'Content-Type: application/json' \
+    -d "{
+        \"amount_mb\": $amount_mb, \
+        \"must_tell_host\": $must_tell_host, \
+        \"deflate_on_oom\": $deflate_on_oom, \
+        \"stats_polling_interval_s\": $polling_interval \
+    }"
+```
+
+To use this, set `socket_location` to the location of the firecracker socket
+(by default, at `/run/firecracker.socket`. Then, set `amount_mb`,
+`must_tell_host`, `deflate_on_oom` and `stats_polling_interval_s` as
+desired: `num_pages` represents the target size of the balloon, and
+`must_tell_host`, `deflate_on_oom` and `stats_polling_interval_s`
+represent the options mentioned before.
+
+To install the balloon via the JSON config file, insert the following JSON
+object into your configuration file:
+
+```
+"balloon": {
+    "amount_mb": $amount_mb,
```

Comment:
The example here is a bit strange as we're giving a JSON file as an example with variables as values.

We could replace them with predefined values, since this is just an example after all.

Reply:
Changed.